pageserver/src/walreceiver.rs

```diff
@@ -648,68 +668,78 @@ impl WalConnectionManager {
+            // check how long time ago did we receive updates from the current SK, switch connections in case it's over the threshold
+            (None, Some((&new_sk_id, _, new_wal_producer_connstr))) => {
+                match self.wal_connection_data.as_ref() {
+                    Some(current_connection) => {
+                        let last_sk_interaction_time =
+                            match current_connection.last_wal_receiver_data.as_ref() {
+                                Some((_, data_submission_time)) => *data_submission_time,
+                                None => current_connection.connection_init_time,
+                            };
+
+                        let now = Utc::now().naive_utc();
+                        match (now - last_sk_interaction_time).to_std() {
+                            Ok(last_interaction) => {
+                                if last_interaction > self.lagging_wal_timeout {
+                                    return Some(NewWalConnectionCandidate {
+                                        safekeeper_id: new_sk_id,
+                                        wal_producer_connstr: new_wal_producer_connstr,
+                                        reason: ReconnectReason::NoWalTimeout {
+                                            last_wal_interaction: last_sk_interaction_time,
+                                            check_time: now,
+                                            threshold: self.lagging_wal_timeout,
+                                        },
+                                    });
+                                }
+                            }
+                            Err(_e) => {
+                                warn!("Last interaction with safekeeper {} happened in the future, ignoring the candidate. Interaction time: {last_sk_interaction_time}, now: {now}", current_connection.safekeeper_id);
+                            }
+                        }
+                        None
+                    }
+                    None => Some(NewWalConnectionCandidate {
+                        safekeeper_id: new_sk_id,
+                        wal_producer_connstr: new_wal_producer_connstr,
+                        reason: ReconnectReason::NoExistingConnection,
+                    }),
                 }
             }
-        }
-    }
-
-    fn reason_to_reconnect(
-        &self,
-        current_connection: &WalConnectionData,
-        new_sk_timeline: &SkTimelineInfo,
-    ) -> Option<ReconnectReason> {
-        let last_sk_interaction_time = match current_connection.last_wal_receiver_data.as_ref() {
-            Some((last_wal_receiver_data, data_submission_time)) => {
-                let new_lsn = new_sk_timeline.commit_lsn?;
-                match new_lsn.0.checked_sub(last_wal_receiver_data.ps_writelsn)
+            // Both current SK got updated via etcd and there's another candidate with suitable Lsn:
```

Comment:
I think this is mostly unreachable code as nothing guarantees candidate sk data and current sk data will be in the same watch batch. Need to check, but in normally I'd expect it deliver single kv in single message.

The way I imagined it is the following: broker maintains map of sk id -> data, and data is timestamped on each new info arrival. Before selecting candidate we remove entries which weren't updated for some time (safekeeper stopped sending data and can be considered dead).